llvm/lib/Passes/PassBuilderPipelines.cpp

```diff
@@ -1828,6 +1828,15 @@ PassBuilder::buildLTODefaultPipeline(OptimizationLevel Level,
     MPM.addPass(PGOIndirectCallPromotion(
         true /* InLTO */, PGOOpt && PGOOpt->Action == PGOOptions::SampleUse));
 
+    // Promoting by-reference arguments to by-value exposes more constants to
+    // IPSCCP.
+    MPM.addPass(
+        createModuleToPostOrderCGSCCPassAdaptor(PostOrderFunctionAttrsPass()));
+    MPM.addPass(
+        createModuleToPostOrderCGSCCPassAdaptor(ArgumentPromotionPass()));
+    MPM.addPass(
+        createModuleToFunctionPassAdaptor(SROAPass(SROAOptions::ModifyCFG)));
```

Comment:
Sorry for not noting this before: Can we make all of these part of a single CGSCC pass manager? I don't think it makes sense to do a separate SCC iteration for FunctionAttrs and ArgPromo.

Reply:
No worries - I have updated to use a single CGSCCPassManager for FunctionAttrs and ArgPromo.

Reply:
You should be able to add SROA to it as well, with a CGSCC to Function pass adaptor. (I think this may also give better results? Otherwise we'll only run SROA after *all* arg promotion, rather than interleaved.)

Reply:
Ah I missed that - I have updated to add SROA to the CGSCCPassManager.